segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/ecs/ECSStorage.java

```diff
@@ -0,0 +1,447 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.ecs;
+
+import com.emc.object.Range;
+import com.emc.object.s3.S3Client;
+import com.emc.object.s3.S3Exception;
+import com.emc.object.s3.S3ObjectMetadata;
+import com.emc.object.s3.bean.AccessControlList;
+import com.emc.object.s3.bean.CanonicalUser;
+import com.emc.object.s3.bean.CopyPartResult;
+import com.emc.object.s3.bean.GetObjectResult;
+import com.emc.object.s3.bean.Grant;
+import com.emc.object.s3.bean.MultipartPartETag;
+import com.emc.object.s3.bean.Permission;
+import com.emc.object.s3.request.CompleteMultipartUploadRequest;
+import com.emc.object.s3.request.CopyPartRequest;
+import com.emc.object.s3.request.PutObjectRequest;
+import com.emc.object.s3.request.SetObjectAclRequest;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.Storage;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.AccessDeniedException;
+import java.time.Duration;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+/**
+ * Storage adapter for extended S3 based Tier2.
+ *
+ * Each segment is represented as a single Object on the underlying storage. As the writes at an offset for an object
+ * file are idempotent, there is no need for locking when a container fails over to another host.
+ * As Pravega does not modify data in Tier2 once written even a contention in writing will cause same data to be
+ * written at the same offset till the time the original host gives up ownership.
+ */
+
+@Slf4j
+public class ECSStorage implements Storage {
+
+    //region members
+
+    private final ECSStorageConfig config;
+    private final ExecutorService executor;
+    private S3Client client = null;
```

Comment:
This is only ever initialized in the tests. When running Pravega this is always null.